railties/lib/rails/generators/rails/app/templates/Dockerfile.tt

```diff
@@ -3,16 +3,16 @@ ARG RUBY_VERSION=<%= Gem.ruby_version %>
 FROM ruby:$RUBY_VERSION
 
 <% if using_node? -%>
-# Install JavaScript dependencies and libvips for Active Storage
-ARG NODE_MAJOR_VERSION=19
-RUN curl -sL https://deb.nodesource.com/setup_$NODE_MAJOR_VERSION.x | bash -
-RUN apt-get update -qq && \
-    apt-get install -y build-essential libvips nodejs && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/* /usr/share/doc /usr/share/man && \
-    npm install -g yarn
+# Install JavaScript dependencies
+ARG NODE_VERSION=<%= dockerfile_node_version %>
+ARG YARN_VERSION=<%= dockerfile_yarn_version %>
+RUN curl https://get.volta.sh | bash
```

Comment:
I'm a bit concerned of using third-party vendors; maybe, we can use official sources? Here is how we do this in [ruby-on-whales](https://github.com/evilmartians/ruby-on-whales/blob/c83cd5f74b83bfc6275179c274619f3f1ed7b7d3/example/.dockerdev/Dockerfile#L38-L47):

```dockerfile
ARG NODE_MAJOR
RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
  --mount=type=cache,target=/var/lib/apt,sharing=locked \
  --mount=type=tmpfs,target=/var/log \
  curl -sL https://deb.nodesource.com/setup_$NODE_MAJOR.x | bash - && \
  DEBIAN_FRONTEND=noninteractive apt-get install -yq --no-install-recommends \
    nodejs

ARG YARN_VERSION
RUN npm install -g yarn@$YARN_VERSION
```

(Ignore `--mount=type=cache`, it worths a separate PR).

Reply:
I would not be happy with that change.

Scenario: At the beginning of the month, after running rigorous regression tests I deploy an application.  Gemfile.lock says stimulus-rails version 1.2.1 so that's the version that is deployed.  yarn.lock says @hotwired/turbo version 7.2.4 so that is the version that is deployed.  At the time of the release Node 19.2.0 is the latest version of node, so that is what is deployed.

Mid month there is a spike in traffic so I deploy a second server.  It gets stimulus-rails 1.2.1, @hotwired/turbo 7.2.4, and.... node 19.3.0.

I realize that dockerfiles aren't completely reproducible.  I may, for example, get a different version of libvips the next time I deploy, but that is a bit more isolated than the node.js runtime upon which a large number of packages depend.

Reply:
Got it (but still don't think upgrading to a new **minor version** of Node could be too harmful; libvips or any other system lib could be).

Another approach (which I've been using in production) is to copy Node files from an official image:

```ruby
FROM node:16-alpine3.16 AS node
FROM ruby:3.1.3-alpine3.16 as main

COPY --from=node /usr/lib /usr/lib
COPY --from=node /usr/local/share /usr/local/share
COPY --from=node /usr/local/lib /usr/local/lib
COPY --from=node /usr/local/include /usr/local/include
COPY --from=node /usr/local/bin /usr/local/bin
COPY --from=node /opt /opt
```

Reply:
I've seen breaking changes in openssh causing upgrade problems for node.js users.

I see you copying from alpine to alpine which should be fine.  Ruby's latest image is based on bullseye.  Some of the previous images are based on buster.  Aligning the node and ruby images would be a bit of a maintenance nightmare.